### PR TITLE
No sandbox const instructions

### PR DESCRIPTION
**Pull request by matthias-blume**
Implements no-sandbox handling of ixx.const instructions.  If the constant refers to an entry in the indirect function table (i.e., if it implements taking a function pointer), then it is rewritten as &f where f is the name of the function in question.

If the constant refers to a memory location (i.e., if it implements taking the address of a global variable) then it is rewritten as &d + addend where d is the name of the relevant data segment and addend is a relative offset within that segment.

This relies on the yet-to-be-implemented simplification of memory handling, where those global variables representing data segments are used directly and are not copied into WASM memory.  (WASM memory itself will go away and become simply memory.)

### CODE REVIEW

File: src/c-writer.cc [hunk 1]
```diff
@@ -3380,6 +3384,71 @@ void CWriter::Write(const ConvertExpr& expr) {
   }
 }
 
+bool CWriter::MaybeWriteNoSandboxFunctionAddress(Offset instruction_offset,
```
Comment (dimitry-) [inline src/c-writer.cc]: nit: maybe TryWrite...

File: src/c-writer.cc [hunk 2]
```diff
@@ -3380,6 +3384,71 @@ void CWriter::Write(const ConvertExpr& expr) {
...
+                                                 bool is64bit) {
+  int index_byte_length = is64bit ? 10 : 5;
```
Comment (dimitry-) [inline src/c-writer.cc:3971]: I am a bit confused by this - can you add a short explanation (or maybe link to spec) here? Does it have something to do with instruction length?

File: src/c-writer.cc [hunk 3]
```diff
@@ -3380,6 +3384,71 @@ void CWriter::Write(const ConvertExpr& expr) {
...
+      instruction_offset - index_byte_length - module_->code_section_base_;
+  auto& function_reloc_map =
+      module_->function_symbol_by_function_pointer_load_offset_;
+  auto function_reloc = function_reloc_map.find(reloc_offset);
+  if (function_reloc == function_reloc_map.end()) {
+    return false;
+  }
+  Index symbol_index = function_reloc->second;
+  Index func_index = module_->function_symbols_.at(symbol_index);
+  std::string& func_name = module_->funcs[func_index]->name;
```
Comment (dimitry-) [inline src/c-writer.cc]: const

File: src/c-writer.cc [hunk 4]
```diff
@@ -3380,6 +3384,71 @@ void CWriter::Write(const ConvertExpr& expr) {
...
+  Write("reinterpret_cast<u", is64bit ? "64" : "32", ">(&",
+        GlobalName(ModuleFieldType::Func, func_name), ")");
+  return true;
+}
+
+bool CWriter::MaybeWriteNoSandboxMemoryAddress(Offset instruction_offset,
+                                               bool is64bit,
+                                               std::string prefix) {
```
Comment (dimitry-) [inline src/c-writer.cc]: const std::string&?

File: src/c-writer.cc [hunk 5]
```diff
@@ -322,6 +322,12 @@ class CWriter {
     Allowed,
   };
 
+  bool MaybeWriteNoSandboxFunctionAddress(Offset instruction_offset,
+                                          bool is64bit);
+  bool MaybeWriteNoSandboxMemoryAddress(Offset instruction_offset,
+                                        bool is64bit,
+                                        std::string prefix = "");
```
Comment (dimitry-) [inline src/c-writer.cc]: The prefix does not seem to be used at least in this PR

File: src/c-writer.cc (hunk 1 above)
Comment (matthias-blume) [reply src/c-writer.cc]: Done.

File: src/c-writer.cc (hunk 2 above)
Comment (matthias-blume) [reply src/c-writer.cc:3971]: Yeah, i32 and i64 varieties of the instructions use 5-byte var[u]int32 and 10-byte var[u]int64 respectively.
I added comments to that effect.

File: src/c-writer.cc (hunk 3 above)
Comment (matthias-blume) [reply src/c-writer.cc]: done (here and in a few other places)

File: src/c-writer.cc (hunk 4 above)
Comment (matthias-blume) [reply src/c-writer.cc]: done

File: src/c-writer.cc (hunk 5 above)
Comment (matthias-blume) [reply src/c-writer.cc]: Yeah, it will be used in the follow-up.  Do you want me to take this out in the meantime?  (Might not be worth the trouble.)

**Comment by matthias-blume** [comment]
ptal

File: include/wabt/ir.h [hunk 6]
```diff
@@ -120,6 +121,7 @@ struct Const {
       case Type::F64: return 2;
       default: WABT_UNREACHABLE;
     }
+    // clang-format on
```
Comment (dimitry-) [inline include/wabt/ir.h]: you can use `git clang-format` (after `git add` but before `git commit`) - it formats only modified code.

File: src/binary-reader-ir.cc [hunk 7]
```diff
@@ -1358,15 +1366,49 @@ Result BinaryReaderIR::EndDataSegmentInitExpr(Index index) {
...
+static Offset GetDataSegmentOffset(const DataSegment* data_segment) {
+  if (data_segment->offset.size() != 1) {
+    return kInvalidOffset;
+  }
+  const auto& expr = *data_segment->offset.begin();
+  if (expr.type() != ExprType::Const) {
+    return kInvalidOffset;
+  }
+  const Const& const_ = cast<ConstExpr>(&expr)->const_;
+  // clang-format off
```
Comment (dimitry-) [inline src/binary-reader-ir.cc]: nit: maybe it is fine to clang-format here?

**Review by dimitry- — APPROVED** [review]

File: include/wabt/ir.h (hunk 6 above)
Comment (matthias-blume) [reply include/wabt/ir.h]: Yeah, I'm aware.  It just happened by accident and I was hoping to avoid this in the future.

Comment (matthias-blume) [reply include/wabt/ir.h]: I removed the clang-format pragmas again.

File: src/binary-reader-ir.cc (hunk 7 above)
Comment (matthias-blume) [reply src/binary-reader-ir.cc]: done

**Pull request merged**